

# NetGPT: Gene regulatory network generation from single-cell transcriptomic sequencing data using prior knowledge-guided diffusion pre-trained model #

**NetGPT will be detailed if the manuscript is accepted!**

**Please write to [zpliu@sdu.edu.cn](mailto:zpliu@sdu.edu.cn) if you have any questions.**

**Please cite our paper if it helps you.**

![workfolw](https://github.com/zpliulab/NetGPT/blob/main/images/logo.png)

## NetGPT User Guide

```NetGPT``` is a tool designed specifically for scRNA-seq data analysis and network modeling. This guide will instruct you on how to train and test models with ```NetGPT```, as well as the requirements for preparing and inputting data.

## Dependencies![Python](https://img.shields.io/badge/python-3.8-blue "Python")

Before you start using NetGPT, please make sure the following dependencies are installed. These packages are necessary for running NetGPT:

#### key packages:
- python==3.8
- torch==2.0.0
- torch-geometric==0.11.4

#### other packages:
- networkx
- joblib
- pickle
- numpy
- pandas
- scikit-learn
- scipy


You can use pip to install these dependent packages, we recommend you use conda to create the environment and install these packages!

```bash
conda create diffusion-NetGPT
conda activate diffusion-NetGPT
conda install python==3.8
```

We recommend using a virtual environment to manage dependencies to avoid any version conflicts.


## Folder and files

- `pre_train/`: The pre-trained NetGPT model provided in this article
- `Cancer_datasets/`: Contains preprocessed results of the HCC/BRCA data used in the manuscript
- `pathway/`: sub-function file containing data preprocessing
- `pathway/simulation/`: synthetic network and gene expression profile generated by SERGIO
- `discrete/`: model-related configuration files
- `discrete/models/`: Multi-head self-attention model
- `denoising_diffusion_pytorch/`: Contains related sub-function files used for training/testing/initializing NetGPT
- `config.py`: Configure hyperparameters for training or testing NetGPT
- `NetGPT.py`: Contains the process framework of NetGPT
- `Download_TF_file.py`: Used to download TF files and gene symbols
- `make_final_net.py`: Integrated voting sub-function
- `Tutorial_running_on_***.py`: A quick tutorial for using NetGPT

Data and pre-trained models can be downloaded from Zenodo!

<div align="center">
  <img src="https://github.com/zpliulab/NetGPT/blob/main/images/network.gif" alt="Schematic diagram of NetGPT generation network" style="width: 200px; height: 100px;"/>
</div>


## Using NetGPT

"Tutorial_running_on_BRCA.py", "Tutorial_running_on_HCC.py" and "Tutorial_running_on_simulation.py" give the data training and testing process in different experimental environments. You can run this file to learn how to use NetGPT!

## A. Training the Model

Depending on your data situation, we divide the training process into two scenarios:

### 1. Input scRNA-seq Gene Expression Profiles with Corresponding GRN

If you have multiple gene expression profiles and their corresponding Gene Regulatory Networks (GRN), you can use these data for model training directly. For specific data format and requirements, please refer to `Supplement S1`.

### 2. Input scRNA-seq Gene Expression Profiles Only

For cases where only gene expression profiles are available, we offer a method to construct a reference network. You can follow the steps described in the manuscript.

**To Start Training**: Once you have prepared the above files, use the following command to start training your model:

```python
from NetGPT import NetGPT
NetGPT.train('Your data file path')
```
## B. Testing the Model

To test a trained NetGPT model, you will need to prepare the following files:

1. **Gene Expression Profile File**: Similar in format to the training file. If true network information is available, it will be evaluated.
2. **Pre-trained NetGPT Model File**: Can be found in our GitHub project's `/result` directory, with files ending in `*.pth`.
3. **PCA Parameter Model File**: This is a PCA parameter file exported based on the training data.

**Optional Parameter**:

- `args.test_pathway`: You can specify a subset of gene sets for network construction, which can be an ID number in the KEGG database or a user-defined list in table format.

**To Start Testing**:

```python
from NetGPT import NetGPT
NetGPT.test('Gene expression profile file path', 'Pre-trained model file path')
```

## Supplementary Notes

### Supplement S1: Data Format Requirements

If your data includes raw data and networks, it can be directly imported in `*.data` format or any pickle saved file. Note, the file should be a dataset containing multiple list-type variables, each with the following structure:

- **'net' variable**: Contains the adjacency matrix of network data, a 0-1 weight matrix (Numpy ndarray). Non-0-1 values can also be loaded, sized cell * cell.
- **'exp' variable**: Contains experimental data in DataFrame format, preprocessed scRNA-seq results (CSV format), sized gene * cell.

### Supplement S2: Data Preprocessing Recommendation

Before inputting sequencing data, we recommend completing matrix completion and quality control. Make sure the input gene expression profiles only contain one cell environment! In addition, although Meta-cell technology and PCA technology have certain contributions in the model, in some cases, users may have to abandon it, for example, only a very small number of cells can be used for analysis.

### Supplementary S3: Other suggestions

Please try to provide NetGPT with a more accurate and specific knowledge base as a controller, which will help obtain a more accurate GRN. It is worth noting that although NetGPT uses Geneformer as a pre-trained gene encoder, other pre-trained models can also be used, even the embedding concatenation of multiple pre-trained models.
